Add Check Support for Java 21 Record Pattern Syntax: IllegalIdentifierName

child of https://github.com/checkstyle/checkstyle/issues/14890
I have read check documentation:https://checkstyle.org/checks/naming/illegalidentifiername.html#IllegalIdentifierName
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words


```
PS D:\CS\test> cat src/Test2.java                                                
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
public class Test2 {
    void test(Object obj) {
        if (!(obj instanceof Rectangle(ColoredPoint record,ColoredPoint yield))) {  // 2 violations
        }
    }

}
PS D:\CS\test> cat config.xml                                                    
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="IllegalIdentifierNameCheck">

        </module>

    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test2.java
Starting audit...
[ERROR] D:\CS\test\src\Test2.java:5:53: Name 'record' must match pattern '(?i)^(?!(record|yield|var|permits|sealed|_)$).+$'. [IllegalIdentifierName]
[ERROR] D:\CS\test\src\Test2.java:5:73: Name 'yield' must match pattern '(?i)^(?!(record|yield|var|permits|sealed|_)$).+$'. [IllegalIdentifierName]
Audit done.
Checkstyle ends with 2 errors.
PS D:\CS\test> 
```

Describe what you expect in detail.

input file

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/IllegalIdentifierNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/IllegalIdentifierNameCheckTest.java
@@ -163,4 +163,21 @@ public class IllegalIdentifierNameCheckTest extends AbstractModuleTestSupport {
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputIllegalIdentifierNameUnnamedVariables.java"), expected);
     }
+
+    @Test
+    public void testIllegalIdentifierNameRecordPattern() throws Exception {
+        final String format = "(?i)^(?!(record|yield|var|permits|sealed)$).+$";
+
+        final String[] expected = {
+            "16:36: " + getCheckMessage(MSG_INVALID_PATTERN, "var", format),
+            "17:47: " + getCheckMessage(MSG_INVALID_PATTERN, "record", format),
+            "17:59: " + getCheckMessage(MSG_INVALID_PATTERN, "yield", format),
+            "17:74: " + getCheckMessage(MSG_INVALID_PATTERN, "sealed", format),
+            "26:28: " + getCheckMessage(MSG_INVALID_PATTERN, "permits", format),
+            "26:41: " + getCheckMessage(MSG_INVALID_PATTERN, "yield", format),
+            "30:39: " + getCheckMessage(MSG_INVALID_PATTERN, "permits", format),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputIllegalIdentifierNameRecordPattern.java"), expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameRecordPattern.java
@@ -1,0 +1,39 @@
+/*
+IllegalIdentifierName
+format = (default)(?i)^(?!(record|yield|var|permits|sealed)$).+$
+tokens = (default)CLASS_DEF, INTERFACE_DEF, ENUM_DEF, ANNOTATION_DEF, ANNOTATION_FIELD_DEF, \
+         PARAMETER_DEF, VARIABLE_DEF, METHOD_DEF, ENUM_CONSTANT_DEF, PATTERN_VARIABLE_DEF, \
+         RECORD_DEF, RECORD_COMPONENT_DEF, LAMBDA
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.naming.illegalidentifiername;
+
+public class InputIllegalIdentifierNameRecordPattern {
+
+    void m(Object o) {
+        if (o instanceof Point(int var, int _)) { } // violation, 'Name 'var' must match *.'
+        if (o instanceof ColorPoint(Point(int record, int yield), String sealed)) { }
+        // 3 violations above:
+        //                    'Name 'record' must match *.'
+        //                    'Name 'yield' must match *.'
+        //                    'Name 'sealed' must match *.'
+    }
+
+    void m2(Object o) {
+        switch (o) {
+            case Point(int permits, int yield): {} break;
+            // 2 violations above:
+            //                    'Name 'permits' must match *.'
+            //                    'Name 'yield' must match *.'
+            case ColorPoint(Point(int permits, int _), String _): {}
+            // violation above, 'Name 'permits' must match *.'
+            default: {}
+        }
+    }
+
+    record Point(int x, int y) { }
+    record ColorPoint(Point p, String color) { }
+
+}

